proof/invariant-abstract/Syscall_AI.thy

```diff
@@ -539,8 +539,8 @@ context Syscall_AI begin
 
 lemma pinv_tcb[wp]:
   "\<And>tptr blocking call can_donate i.
-    \<lbrace>\<lambda>s. invs s \<and> st_tcb_at active tptr s \<and> ct_active s \<and> valid_invocation i s \<and>
-         scheduler_action s = resume_cur_thread\<rbrace>
+    \<lbrace>\<lambda>s. invs s \<and> st_tcb_at active tptr s \<and> ct_active s \<and> valid_invocation i s
+         \<and> bound_sc_tcb_at (\<lambda>a. \<exists>y. a = Some y) (cur_thread s) s\<rbrace>
```

Comment:
`~= None` here as well.  Do we not have an abbreviation for just "is bound"? Might have been a different proof that I'm remembering.

Reply:
We do have `bound`, which is an abbreviation for `_ ~= None`. But when I rewrite the rules and statements in preconditions in this way, I think the simplifier will turn this into the exists form, and then I can't use my rules (I have to use `not_None_eq` to get it back into a form that can be used). Any thoughts?

Reply:
You can use `your_rule[simplified]` and then it should match, I think.

Reply:
Thanks Miki, that did wok. It's a bit unfortunate that you have to take a rule already in the `wp` set and then mention it explicity with the simplified attribute. If this is the best way to go, that's fine, but I'd be curious if there is another way

Reply:
I have next to no experience with simp sets or building something like `wpsimp`, but if `wpsimp` is basically alternating `wp` and `simp`, doesn't it make sense for `wpsimp` and friends to be trying to work both plain and simplified versions of rules? Maybe this would kill performance

Reply:
A few options:
- change `bound` to the current simp-normal-form, which is `Ex .. = Some`. This doesn't look nice in the definition of `bound`, but would probably be the most stable one.
- change the attribute at the lemma declaration from `[wp]` to `[simplified, wp]`. This would work, but is not generally nice style, because now the lemma that is written down is not the actual lemma statement.
- remove the rule `not_None_eq` from the simp set where we need these lemmas to apply. This is not very stable in general, and sometimes you might want the rule present because you want other `~= None` rewritten, but I reckon it would work here.

Reply:
I'd prefer the first option. Is this something we should be doing on master? If so, I might just make a JIRA issue for this. If just for RT, and others like the idea too, I might make this change first and then rebase this PR on top

Reply:
It's Ok to do on `rt` first, but it would be nice to make the `bound` part a separate commit if possible, so that we can apply it to master later as well.

Reply:
OK, I'll do it on `rt` first and make it separate commit(s)